components/clp-package-utils/clp_package_utils/controller.py

```diff
@@ -638,8 +650,74 @@ class DockerComposeController(BaseController):
     Controller for orchestrating CLP components using Docker Compose.
     """
 
-    def __init__(self, clp_config: CLPConfig, instance_id: str) -> None:
+    def __init__(
+        self,
+        clp_config: CLPConfig,
+        instance_id: str,
+        target: DeploymentTarget,
+        num_workers: Optional[int]
+    ) -> None:
         self._project_name = f"clp-package-{instance_id}"
+
+        self._target = target
+        self._is_existing_project_allowed = target != DeploymentTarget.ALL
+        self._num_workers = num_workers or self._get_num_workers()
+
+        self._is_mcp_enabled = clp_config.mcp_server is not None
+        self._is_legacy_search_enabled = clp_config.package.query_engine != QueryEngine.PRESTO
+
+        if target in (
+                DeploymentTarget.QUERY_WORKER,
+                DeploymentTarget.REDUCER,
+        ) and not self._is_legacy_search_enabled:
+            raise ValueError(
+                "Legacy search components (query-worker/reducer) cannot be launched when the "
+                "query engine is set to Presto."
+            )
+        if target == DeploymentTarget.MCP and not self._is_mcp_enabled:
+            raise ValueError(
+                "The MCP server is not configured in the CLP package configuration, so it cannot "
+                "be launched."
+            )
+
+        # Controllers
+        self._launch_database = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_redis = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_queue = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_results_cache = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_compression_scheduler = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_query_scheduler = target in (DeploymentTarget.ALL, DeploymentTarget.CONTROLLER)
+        self._launch_garbage_collector = target in (DeploymentTarget.ALL,
+                                                    DeploymentTarget.CONTROLLER) and is_retention_period_configured(clp_config)
+
+        # Workers
+        self._launch_compression_worker = target in (DeploymentTarget.ALL, DeploymentTarget.COMPRESSION_WORKER)
+        self._launch_query_worker = target in (DeploymentTarget.ALL, DeploymentTarget.QUERY_WORKER)
+        self._launch_reducer = target in (DeploymentTarget.ALL, DeploymentTarget.REDUCER)
+
+        # Clients
+        self._launch_webui = target in (DeploymentTarget.ALL, DeploymentTarget.WEBUI)
+        self._launch_mcp = (target in (DeploymentTarget.ALL,DeploymentTarget.MCP) and self._is_mcp_enabled)
```

Comment:
some of (, if not most of) those computed flags are unused. i was trying to use those to control whether the setups methods should be run or not, but realized we shouldn't disable any setup functions as they may setup dependencies for the other components.

e.g. the db's setup method should not be disabled if we want to run the webui, or the webui might not be able to read the credential env vars